### PR TITLE
Fix 944: Fixed edge case of cli argument duplication for `%sqlcmd` magic

### LINKED ISSUE #944: fix cli duplicated arg validation edge case
see:  https://github.com/ploomber/jupysql/pull/908#issuecomment-1816197125

### PR DESCRIPTION
**Pull request by nooodle-soup**
## Describe your changes
1. Added logic to disable use of more than one delete commands in `%sqlcmd snippets`. No longer allows any combination of `-d`, `--delete`, `-D`, `--delete-force`, `-A` and `--delete-force-all` for the %sqlcmd snippets magic.
2. Added test to reflect changes in the `check_duplicate_arguments` function.

## Issue number

Closes #944 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--946.org.readthedocs.build/en/946/

### CODE REVIEW

File: src/sql/util.py [hunk 1]
```diff
@@ -251,17 +266,26 @@ def check_duplicate_arguments(
     else:
         duplicates_error = ""
 
-    if alias_pairs_present:
-        arg_list = sorted([" or ".join(pair) for pair in alias_pairs_present])
+    if alias_pairs_present and not len(delete_arg_present) > 1:
```
Comment (neelasha23) [inline src/sql/util.py:269]: The condition can be simplified for readability: `len(delete_arg_present) <= 1`

File: src/tests/conftest.py [hunk 2]
```diff
@@ -20,30 +20,49 @@
...
+            delete_arg_list = [
+                "-d",
+                "-D",
+                "-A",
+                "--delete",
+                "--delete-force",
+                "--delete-force-all",
+            ]
+            delete_args_present = set()
+            for arg in args:
```
Comment (neelasha23) [inline src/tests/conftest.py:55]: this code seems repeated from the util function. I didn't understand why we need it here

Comment (nooodle-soup) [reply src/tests/conftest.py:55]: This function is a message generation factory so it generates the error message based on a set of inputs. As the edge case we're handling here is very specific, the logic for message generation is almost the same as that from the util function.

File: src/tests/conftest.py [hunk 3]
```diff
@@ -20,30 +20,49 @@
 
 @pytest.fixture
 def check_duplicate_message_factory():
-    def _generate_error_message(cmd, args, aliases=None):
+    def _generate_error_message(cmd, args, aliases=None, delete_present=False):
```
[the remark this answers is not in the document]
Comment (nooodle-soup) [reply src/tests/conftest.py:23]: If there is any of the 6 `delete` arguments in the magic the user is trying to run, `delete_present` would be `True`. I'm using it as a flag to control message generation through the factory.

File: src/sql/util.py (hunk 1 above)
Comment (nooodle-soup) [reply src/sql/util.py]: @neelasha23 Fixed [here](https://github.com/ploomber/jupysql/pull/946/files#diff-653c15e6161c97f356be04e072838ed15014af275c06019a7692ab727fa44b97R269)

File: src/tests/conftest.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nooodle-soup) [reply src/tests/conftest.py]: I've changed the name of the flag to `delete_option_used` for ease of understanding [here](https://github.com/ploomber/jupysql/pull/946/files#diff-d693b2faa093742e7045e9d6606e499d5ff59a8c9625968286d5abf230c28e35R23). @neelasha23

File: src/tests/conftest.py [hunk 4]
```diff
@@ -20,30 +20,49 @@
...
+                f"{', '.join(sorted(duplicates))}. "
             )
-            if aliases:
-                error_message += " "
 
-        if aliases:
+        if aliases and not delete_present:
             alias_list = []
             for pair in sorted(aliases):
                 print(pair[0], pair[1])
```
[the remark this answers is not in the document]
Comment (nooodle-soup) [reply src/tests/conftest.py]: Removed in [this](https://github.com/ploomber/jupysql/pull/946/commits/134a9fdd57a17232a956529a7cf1b656e7cc323c) commit. @neelasha23

File: src/tests/conftest.py (hunk 2 above)
Comment (neelasha23) [reply src/tests/conftest.py:54]: Cant we just mention all the error messages in `parametrize` as in the other tests? There will be multiple cases but atleast the readability and debugging improves.

Comment (neelasha23) [reply src/tests/conftest.py:54]: Also, the error message suffix/ prefix can be mentioned as string constants and reused in the tests. so that if the message changes the tests also reflect the same.

**Comment by neelasha23** [comment]
Please close this PR @nooodle-soup

**Comment by nooodle-soup** [comment]
Closing as per discussions in #944
